### DISCUSSION TITLE
"Operation not permitted" creating directory in $(HOMEBREW_PREFIX)/lib

### DISCUSSION THREAD

**Discussion by rexx-org** [Homebrew/discussions#3515] [Tap maintenance and Homebrew development]
I am attempting to update the regina-rexx formula to support installing add-on packages for Regina. Add-on packages are dynamically loaded .dylibs which enhance the base Rexx language. Think of them as like Ruby Gems.

On my M1 Mac I have a directory: /opt/homebrew/lib/ruby/gems. I am wanting to replicate this directory structure so that Regina add-on packages can link their .dylibs into /opt/homebrew/lib/regina-rexx/addons.

My problem is that when I try and create /opt/homebrew/lib/regina/rexx-addons when I build from source I get the following error:
./install-sh -m 755 -c ./regutil/regutil.rexx  /opt/homebrew/Cellar/regina-rexx/3.9.5/share/regina-rexx/examples/regutil.rexx
./install-sh -m 755 -d /opt/homebrew/Cellar/regina-rexx/3.9.5/lib
./install-sh -m 755 -d /opt/homebrew/lib/regina-rexx/addons
mkdir: /opt/homebrew/lib/regina-rexx: Operation not permitted
mkdir: /opt/homebrew/lib/regina-rexx: No such file or directory
make: *** [installmaclib] Error 1

The "brew install --build-from-source regina-rexx" is run as user "mark". I can manually "mkdir /opt/homebrew/lib/regina-rexx"
Output from:
ls -ld@eO /opt/homebrew/lib                                                                                                                                                                                      
is:
drwxrwxr-x  482 mark  admin  - 15424 10 Aug 09:13 /opt/homebrew/lib

My question is how do I create the regina-rexx/addons directory in /opt/homebrew/lib as part of the regina-rexx formula?

**Comment by carlocab** [comment]
Formulae are not allowed to write directly into `HOMEBREW_PREFIX` (in your case, this is `/opt/homebrew`). Instead, they have their own private prefix, which looks something like `$HOMEBREW_PREFIX/Cellar/regina-rexx/x.y.z`. This path is accessible in the formula via the variable `prefix`. `brew` handles symlinking the contents of this prefix into `HOMEBREW_PREFIX`.

See https://docs.brew.sh/Formula-Cookbook and in particular https://docs.brew.sh/Formula-Cookbook#variables-for-directory-locations.

**Reply by rexx-org** [reply]
Yes I understand that the package's cellar has its own prefix. The files that Regina uses internally are installed in this prefix. The Regina package has an add-on; libregutil.dylib, that is installed in the Cellar's internal prefix. I then want that file to be linked into the "shared" location /opt/homebrew/lib/regina-rexx/addons; that is what I am trying to do when I encounter the reported error.

Let's take pkgconfig as another example.  There is /opt/homebrew/lib/pkgconfig where packages other than pkgconfig link their .pc files into this "shared" directory. /opt/homebrew/lib/pkgconfig is not pkgconfig's internal prefix. Presumably pkgconfig creates /opt/homebrew/lib/pkgconfig so that any other package that requires a .pc file to be installed can be linked into this directory. If pkgconfig, and ruby, and perl5 and python can create a directory in /opt/homebrew/lib then why is my package prohibited from doing the same?

**Reply by carlocab** [reply]
Well, your formula is trying to do:
```
./install-sh -m 755 -d /opt/homebrew/lib/regina-rexx/addons
```
It shouldn't do that. Again, it should not be trying to touch anything outside its own prefix. What it should be doing is something like
```
./install-sh -m 755 -d /opt/homebrew/Cellar/regina-rexx/3.9.5/lib/regina-rexx/addons
```
and let `brew` handle linking that into `/opt/homebrew/lib` afterward.

**Reply by rexx-org** [reply]
Thanks for the information carlocab. I changed the install to above. I now have the following relevant entries in /opt/homebrew/lib:
drwxr-xr-x    3 mark  admin    96 30 May 08:22 perl5
drwxr-xr-x  131 mark  admin  4192 10 Aug 11:18 pkgconfig
drwxr-xr-x    3 mark  admin    96 21 Jan  2022 python3.10
drwxr-xr-x    3 mark  admin    96 14 Oct  2021 python3.9
lrwxr-xr-x    1 mark  admin    43 10 Aug 11:18 regina-rexx -> ../Cellar/regina-rexx/3.9.5/lib/regina-rexx
drwxr-xr-x    3 mark  admin    96 30 May 08:21 ruby

So /opt/homebrew/regina-rexx is a symbolic link into Regina's Cellar, which is version dependant.  Other packages are not allowed to write into another package's Cellar, so this will not work.
Also when I upgrade Regina to say 3.9.6 Cellar/regina-rexx/3.9.5/lib/regina-rexx will no longer exist and any symbolic links from other packages (which can't be created anyway) will be left dangling.
Again I emphasise the difference with /opt/homebrew/lib/regina-rexx (a symbolic link into a Cellar) and /opt/homebrew/pkgconfig (a directory; not linked into a Cellar).
How is the directory /opt/homebrew/lib/pkgconfig created?

**Reply by gromgit** [reply]
I believe what you're seeing is because there's currently only one formula installed on your machine that has a `lib/regina-rexx/addons` directory, viz. `regina-rexx` itself. (If you had only one formula installed that had a `lib/pkgconfig` subdirectory, you'd see `/opt/homebrew/lib/pkgconfig` also be a single symlink to that subdirectory, instead of being a directory itself.) 

Try installing another formula that has its own `lib/regina-rexx/addons` stuff. You'll likely see that, as in the `pkgconfig` case, Homebrew will automagically convert your current `/opt/homebrew/lib/regina-rexx` symlink into a proper directory hierarchy, so that the multiple files in each formula's `lib/regina-rexx/addons` can be properly symlinked in the correct location, just like `pkgconfig`.

**Reply by rexx-org** [reply]
Thank you Adrian. I would never have guessed that behaviour. I will do as you suggest and wait for the magic to happen!

**Reply by rexx-org** [reply]
Well the magic happened! A few things to tweak, but I have achieved the behaviour I required. Thanks again Adrian!
